sktime/datatypes/_vectorize.py

```diff
@@ -310,6 +309,118 @@ def _get_item_indexer(self, i: int, X=None):
             col_ind = [col_ind]
         return row_ind, col_ind
 
+    def __iter__(self):
+        """Iterate over all instances.
+
+        Returns
+        -------
+        An iterator over all instances
+        """
+        return (
+            group
+            for _, _, group in self.items(
+                iterate_as=self.iterate_as, iterate_cols=self.iterate_cols
+            )
+        )
+
+    def items(self, iterate_as=None, iterate_cols=None):
+        """Iterate over (group name, column name, instance) tuples.
+
+        Iteration order is "depth first" with columns being the branches and groups the
+        root nodes. Groups are ordered by group keys.
+
+        Row name is null if iterate_as is the same as scitype of data.
+        Col name is null if iterate_cols is False.
+
+        Parameters
+        ----------
+        iterate_as : str ("Series", "Panel"), optional, default=self.iterate_as
+            scitype of the iteration
+            for instance, if X is Panel and iterate_as is "Series"
+                then the class will iterate over individual Series in X
+            or, if X is Hierarchical and iterate_as is "Panel"
+                then the class will iterate over individual Panels in X
+                    (Panel = flat/non-hierarchical collection of Series)
+        iterate_cols : boolean, optional, default=self.iterate_cols
+            whether to iterate over columns, if true, the class will iterate over
+            columns
+
+        Returns
+        -------
+        An iterator over all (row name, column name, instance) tuples.
```

Comment:
this is not 100% correct - it doesn't return an iterator, but *is* a generator, no? Also, is this not going to get us into trouble when we, say, want to iterate through it twice?

Reply:
Update, note to myself: I think no, because each call a new generator is returned.

Reply:
To the cite python docs:
> Python’s [generator](https://docs.python.org/3/glossary.html#term-generator)s provide a convenient way to implement the iterator protocol. If a container object’s __iter__() method is implemented as a generator, it will automatically return an iterator object (technically, a generator object) supplying the __iter__() and [__next__()](https://docs.python.org/3/reference/expressions.html#generator.__next__) methods. More information about generators can be found in [the documentation for the yield expression](https://docs.python.org/3/reference/expressions.html#yieldexpr).

https://docs.python.org/3/library/stdtypes.html#generator-types

And yes, each time a new generator is returned.

Reply:
well, we should at least update the docstring then for it to be correct. It returns a generator, not an iterator.

Reply:
I hope I didn't just end up confusing myself, so please correct me if I'm wrong.

Reply:
Technically, you are correct. But a generator [behaves like an iterator](https://wiki.python.org/moin/Generators) and I'm my opinion, this is an implementation detail, a consumer of the method shouldn't care about. Don't you think so?

Reply:
I do, but docstrings should be correct up to the type. A consumer will not care in most cases, but I think it's good practice to be technically correct.

How about saying sth like `generator, iterates over all (row name, column name, instance) tuples. The i-th element returned by the generator is ...`

Reply:
ok, I'll merge it, and we can talk the docstring separately